dds/DCPS/RTPS/Sedp.cpp

```diff
@@ -3671,25 +3682,92 @@ Sedp::TypeLookupRequestReader::process_get_types_request(const XTypes::TypeLooku
 }
 
 DDS::ReturnCode_t
-Sedp::TypeLookupReplyReader::process_tl_reply(DCPS::Serializer& ser)
+Sedp::TypeLookupRequestReader::process_get_dependencies_request(const XTypes::TypeLookup_Request& request,
+  XTypes::TypeLookup_Reply& reply)
+{
+  // TODO(sonndinh): Is sending a complete set of dependencies ok in case it is big?
+  // How is continuation_point set?
+  sedp_.type_lookup_service_->get_type_dependencies(request.data.getTypeDependencies.type_ids,
+    reply.data.getTypeDependencies.result.dependent_typeids);
+  if (reply.data.getTypeDependencies.result.dependent_typeids.length() > 0) {
+    reply.data.kind = XTypes::TypeLookup_getDependencies_HashId;
+    reply.data.getTypeDependencies.return_code = DDS::RETCODE_OK;
+    reply.header.related_request_id = request.header.request_id;
+    return DDS::RETCODE_OK;
+  }
+  return DDS::RETCODE_NO_DATA;
+}
+
+DDS::ReturnCode_t
+Sedp::TypeLookupReplyReader::process_type_lookup_reply(const DCPS::ReceivedDataSample& sample,
+                                                       DCPS::Serializer& ser)
 {
   XTypes::TypeLookup_Reply type_lookup_reply;
   if (!(ser >> type_lookup_reply)) {
-    ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: Sedp::TypeLookupReplyReader::process_tl_reply - ")
-              ACE_TEXT("failed to deserialize type lookup reply\n")));
+    ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: Sedp::TypeLookupReplyReader::process_type_lookup_reply - ")
+               ACE_TEXT("failed to deserialize type lookup reply\n")));
     return DDS::RETCODE_ERROR;
   }
 
-  if (type_lookup_reply.data.getTypes.result.types.length() > 0) {
-    sedp_.type_lookup_service_->add_type_objects_to_cache(type_lookup_reply.data.getTypes.result.types);
+  DDS::ReturnCode_t retcode;
+  switch (type_lookup_reply.data.kind) {
+  case XTypes::TypeLookup_getTypes_HashId:
+    retcode = process_get_types_reply(type_lookup_reply);
+    break;
+  case XTypes::TypeLookup_getDependencies_HashId:
+    retcode = process_get_dependencies_reply(sample, type_lookup_reply);
+    break;
+  default:
+    return DDS::RETCODE_UNSUPPORTED;
+  }
+
+  if (DDS::RETCODE_OK == retcode) {
     DCPS::SequenceNumber rpc_sequence;
     rpc_sequence.setValue(type_lookup_reply.header.related_request_id.sequence_number.high,
-      type_lookup_reply.header.related_request_id.sequence_number.low);
+                          type_lookup_reply.header.related_request_id.sequence_number.low);
+    // TODO(sonndinh): Is this a correct place to call match_continue?
```

Comment:
We can't call match_continue() until we receive both Type Objects and ALL Dependencies (which becomes even more interesting, as getting dependencies may result in subsequent request for the Type Objects). Somehow we should keep track (counter? - as a suggestion) if any Type Lookup Service requests are outstanding and proceed with match_continue(), only if the counter is 0.

Reply:
This is handled now.